### ISSUE TITLE
Regression: discovery for @Nested test classes with same superclass no longer works

### ISSUE THREAD

**Issue by mbush8** [junit-team/junit5#1954]
**Version 5.5.0**

Given the following:

```java
    abstract class AbstractClassWithNestedTest {

        abstract String expectedValue();

        @Nested
        class AbstractNested {

            @Test
            void abstractClassAssert1() {
                Assertions.assertEquals(expectedValue(), "pass");
            }
        }
    }
```

and

```java
class MainTest {

    @Nested
    class Inner1  extends AbstractClassWithNestedTest {

        @Override
        String expectedValue() {
            return "pass";
        }
    }

    @Nested
    class Inner2 extends AbstractClassWithNestedTest {

        @Override
        String expectedValue() {
            return "fail";
        }
    }
}
```

**Expected Behavior:** `abstractClassAssert1()` to execute twice; once as instance of `Inner1` with a passing test and once as instance of `Inner2` with a failing test.

**Actual Behavior:** `abstractClassAssert1()` executes only once as instance of `Inner1` with a passing test.

If I remove class `Inner1` then `abstractClassAssert1()` executes only once as instance of `Inner2` and test fails.

If I declare class `Inner2` before class `Inner1` then `abstractClassAssert1()` executes  only once as instance of `Inner2` and test fails.

If I remove class `AbstractNested` and declare `abstractClassAssert1()` inside of class `AbstractClassWithNestedTest` then `abstractClassAssert1()` executes twice; once as instance of `Inner1` with a passing test and once as instance of `Inner2` with a failing test.

**Comment by sbrannen** [comment]
> **Actual Behavior:** `abstractClassAssert1()` executes only once as instance of `Inner1` with a passing test.

That's rather interesting, because I see the opposite behavior: `abstractClassAssert1()` executes only once as instance of `Inner2` with a failing test.

But I think this is really just what the IDE is reporting. For example, I'm using Eclipse.

And I assume you're likely using something different such as IntelliJ -- right?

**Comment by sormuras** [comment]
What is the output when launched on the console? Via https://junit.org/junit5/docs/current/user-guide/#running-tests-console-launcher or Gradle or Maven/Surefire?

**Comment by sbrannen** [comment]
I've already got a Test Kit example in the works showing that only one of the nested test classes is executing. So I don't think the console launcher will work any differently. 😉

**Comment by sbrannen** [comment]
For example, the following fails since only 4 containers are found instead of 6.

```java
class TestKitTests {

	@Test
	void executeTests() {
		EngineExecutionResults results = EngineTestKit.engine("junit-jupiter")//
				.selectors(selectClass(MainTest.class))//
				.execute();

		results.allEvents().debug(System.out);
		results.containerEvents().assertStatistics(stats -> stats.started(6));
		results.testEvents().assertStatistics(stats -> stats.started(2));
	}

}
```

**Comment by sbrannen** [comment]
And here's the debug output for all events.

```
All Events:
	Event [type = STARTED, testDescriptor = JupiterEngineDescriptor: [engine:junit-jupiter], timestamp = 2019-07-10T15:21:39.571899Z, payload = null]
	Event [type = STARTED, testDescriptor = ClassTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest], timestamp = 2019-07-10T15:21:39.572543Z, payload = null]
	Event [type = STARTED, testDescriptor = NestedClassTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest]/[nested-class:Inner2], timestamp = 2019-07-10T15:21:39.573665Z, payload = null]
	Event [type = STARTED, testDescriptor = NestedClassTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest]/[nested-class:Inner2]/[nested-class:AbstractNested], timestamp = 2019-07-10T15:21:39.574516Z, payload = null]
	Event [type = STARTED, testDescriptor = TestMethodTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest]/[nested-class:Inner2]/[nested-class:AbstractNested]/[method:abstractClassAssert1()], timestamp = 2019-07-10T15:21:39.575301Z, payload = null]
	Event [type = FINISHED, testDescriptor = TestMethodTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest]/[nested-class:Inner2]/[nested-class:AbstractNested]/[method:abstractClassAssert1()], timestamp = 2019-07-10T15:21:39.594138Z, payload = TestExecutionResult [status = FAILED, throwable = org.opentest4j.AssertionFailedError: expected: <fail> but was: <pass>]]
	Event [type = FINISHED, testDescriptor = NestedClassTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest]/[nested-class:Inner2]/[nested-class:AbstractNested], timestamp = 2019-07-10T15:21:39.594992Z, payload = TestExecutionResult [status = SUCCESSFUL, throwable = null]]
	Event [type = FINISHED, testDescriptor = NestedClassTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest]/[nested-class:Inner2], timestamp = 2019-07-10T15:21:39.595028Z, payload = TestExecutionResult [status = SUCCESSFUL, throwable = null]]
	Event [type = FINISHED, testDescriptor = ClassTestDescriptor: [engine:junit-jupiter]/[class:org.junit.platform.MainTest], timestamp = 2019-07-10T15:21:39.595066Z, payload = TestExecutionResult [status = SUCCESSFUL, throwable = null]]
	Event [type = FINISHED, testDescriptor = JupiterEngineDescriptor: [engine:junit-jupiter], timestamp = 2019-07-10T15:21:39.595241Z, payload = TestExecutionResult [status = SUCCESSFUL, throwable = null]]
```

Class `Inner2` doesn't show up anywhere.

**Comment by sbrannen** [comment]
Tentatively slated for 5.6 M1 since this appears to be a bug.

**Comment by sbrannen** [comment]
FWIW, it doesn't matter that the parent class is `abstract`. The test kit test fails in the same manner with the following.

```java
	static class AbstractClassWithNestedTest {

		String expectedValue() {
			return "foo";
		}

		@Nested
		class AbstractNested {

			@Test
			void abstractClassAssert1() {
				Assertions.assertEquals(expectedValue(), "pass");
			}
		}
	}
```

**Comment by sbrannen** [comment]
@marcphilipp, here's everything in one source file to simplify matters.

```java
import static org.junit.platform.engine.discovery.DiscoverySelectors.selectClass;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.platform.testkit.engine.EngineExecutionResults;
import org.junit.platform.testkit.engine.EngineTestKit;

public class MainTest {

	@Nested
	class Inner1 extends AbstractClassWithNestedTest {

		@Override
		String expectedValue() {
			return "pass";
		}
	}

	@Nested
	class Inner2 extends AbstractClassWithNestedTest {

		@Override
		String expectedValue() {
			return "fail";
		}
	}

	static abstract class AbstractClassWithNestedTest {

		abstract String expectedValue();

		@Nested
		class AbstractNested {

			@Test
			void abstractClassAssert1() {
				Assertions.assertEquals(expectedValue(), "pass");
			}
		}
	}

	static class TestKitTests {

		@Test
		void executeTests() {
			EngineExecutionResults results = EngineTestKit.engine("junit-jupiter")//
					.selectors(selectClass(MainTest.class))//
					.execute();

			results.allEvents().debug(System.out);
			results.containerEvents().assertStatistics(stats -> stats.started(6));
			results.testEvents().assertStatistics(stats -> stats.started(2));
		}

	}
}
```

**Comment by sbrannen** [comment]
**Update**: this passes against the `4.3.x` branch, so it's a regression in 5.5.

```java
	static class TestKitTests {

		@Test
		void executeTests() {
			EngineExecutionResults results = EngineTestKit.engine("junit-jupiter")//
					.selectors(selectClass(MainTest.class))//
					.execute();

			results.all().debug(System.out);
			results.containers().assertStatistics(stats -> stats.started(6));
			results.tests().assertStatistics(stats -> stats.started(2));
		}

	}
```

Using _old_ test kit event filter method names.

**Comment by marcphilipp** [comment]
@mbush8 If you have time, it would be great if you could verify the fix using the latest 5.5.1-SNAPSHOT in your project.
